### ISSUE TITLE
structure of "other stuff"

### ISSUE THREAD

**Issue by crotwell** [iris-edu/mseed3-evaluation#1]
Within the old strawman, there was a transition from blockettes to just having essential stuff in the mseed header, and with an "other stuff" section for things that might be useful but were not absolutely essential. The format was just simple ~ delimited strings, allowing things like name=value~

I feel like there should be a more structured way of doing this. Two ideas, first is messagepack. It is binary, pretty simple and would allow objects/arrays in addition to name-value. Lots of language bindings and writing a writer or parser feels easy enough to me.
http://msgpack.org/

Second idea, as I mentioned at the meeting, would be to just use json. At least you get a lot of structure and self description, plus separation between numbers and non-numbers.
http://json.org/

I prefer messagepack, given that mseed is binary and so maybe the other stuff should be as well, but not strongly. Json is pretty nice and if we are going to have strings anyway, the extra overhead seems small.

Thoughts?

**Comment by crotwell** [comment]
Extremely simple example of creating cbor, uses json as the input. Shows the bytes and the length. Not really that useful in itself, but just letting you know it as might be useful basis for more experimenting with cbor.

https://github.com/iris-edu/mseed3-evaluation/tree/master/Crotwell/extraHeaders/cbor

**Comment by chad-iris** [comment]
Just ran across this.

[CBOR](https://tools.ietf.org/html/rfc7049) there is a schema definition on the standards track called [CDDL](https://tools.ietf.org/html/draft-greevenbosch-appsawg-cbor-cddl-11).

Abstract

> This document proposes a notational convention to express CBOR data structures (RFC 7049).  Its main goal is to provide an easy and  unambiguous way to express structures for protocol messages and data formats that use CBOR.

Small example is here: http://cbor.io/tools.html

**Comment by crotwell** [comment]
Yep, I saw that and even started to translate the json-schema to cddl, but got distracted. Maybe will give it another try. We probably should explore using cbor at least as fully as we explore json.

Only qualm I have is the date on the CDDL Internet-Draft is July 03, 2017. So this is pretty new. No idea of what the likelihood is for changes before (or if) it gets to be an RFC. My guess is that as long as we only use the basics, it should be fine. And the FDSN process is likely to take longer than the IETF, so plenty of chance to catch any modifications to it.

Is the general consensus that CBOR is the most promising of the binary json formats, just in terms of where to devote our limited time and energy?

**Comment by chad-iris** [comment]
> Is the general consensus that CBOR is the most promising of the binary json formats, just in terms of where to devote our limited time and energy?

It is for me, the more I look at it the more I like it.  I will be evaluating both a hybrid binary-CBOR and a full CBOR implementation.

@andres-h  is [evaluating Protouf](https://github.com/iris-edu/mseed3-evaluation/issues/14#issuecomment-315721078), which is good to have experience with.

If we want to go with one of these binary encodings, for extras or "full record", we'll need to decide between the two.  I'm leaning towards CBOR because one of there primary goals is "CBOR is defined in an Internet Standards Document, RFC 7049. The format has been designed to be stable for decades.", whereas I do not think that was ever a goal for Protobuf.  Not that it couldn't be used, just don't know.

**Comment by andres-h** [comment]
In which aspects is CDDL better than JSON schema? If CBOR is just a binary representation of JSON, one could use JSON schema for CBOR and vice versa.

> The format has been designed to be stable for decades.", whereas I do not think that was ever a goal for Protobuf.

That is a terrible misunderstanding. AFAIK there have never been any changes in the Protobuf encoding since it was released to the public (2008), and it is designed to be fully forward- and backward-compatible.

As I wrote before, Protobuf encoding should not be mixed with Protobuf software. It is like mixing miniSEED and libmseed (eg., "miniSEED is not stable, because the API of libmseed may change").

**Comment by crotwell** [comment]
Did some more reading, CDDL does not appear to be that useful for extra headers. We could define a CDDL "schema" for the fdsn standard parts, but there does not seem to be a way to do the equivalent of "additionalProperties" as in json-schema. So there isn't an easy way to simply have other as yet undefined key-value pairs but also validate with the schema. Or at least I didn't see it.

This may not be a killer issue, as an extra header spec could exist in CBOR without validation via CDDL, but it is not as simple as the json-schema case.

Also, stumbled into some more information on the messagepack cbor relationship. Looks like cbor started as a hostile fork of messagepack, some bad feelings on both sides with a rush to get RFC status. I don't know that is a reason to be for or against it, but not all warm and fuzzy.  I am also rethinking the RFC issue a bit. An RFC is desirable if it reflects a widespread consensus and a likelyhood of longevity, but there are plenty of RFC's that were approved and then fell out of use. In other words, the spec has to have merit on its own regardless of if an RFC exists. 

Just some numbers of tagged messages on stackoverflow:

- cbor 15
- messagepack 270
- jsonb 560
- protobuf 3257
- json 205,408



Some links:
https://github.com/msgpack/msgpack/issues/129
https://www.ietf.org/mail-archive/web/json/current/msg00195.html

**Comment by chad-iris** [comment]
@andres-h 
> In which aspects is CDDL better than JSON schema? If CBOR is just a binary representation of JSON, one could use JSON schema for CBOR and vice versa.

I don't know exactly yet.  One could guess that if JSON Schema were a perfect fit they would never have made CDDL.  CDDL is directly targeting CBOR and handles it's types, whereas converting CBOR to JSON is not a direct mapping and there is only non-normative advice to do that operation:
https://tools.ietf.org/html/rfc7049#section-4.  So I'm speculating that it's a "better fit".

>> The format has been designed to be stable for decades.", whereas I do not think that was ever a goal for Protobuf.

> That is a terrible misunderstanding. AFAIK there have never been any changes in the Protobuf encoding since it was released to the public (2008), and it is designed to be fully forward- and backward-compatible.

Please do not exaggerate what I wrote.  I said not a goal of that encoding and I stand by that.  I did not say it was not stable or would not be stable.

**Comment by crotwell** [comment]
@andres-h 

> In which aspects is CDDL better than JSON schema? If CBOR is just a binary representation of JSON, one could use JSON schema for CBOR and vice versa.

CBOR is not just a binary json. It uses similar ideas, but has more structure. For example, json only has "number" but cbor has floats and ints, etc. Similar to the idea you wouldn't use xml-schema to validate json, the schema language and the format need to be tightly related.

**Comment by chad-iris** [comment]
> Did some more reading, CDDL does not appear to be that useful for extra headers. 

Bummer. I don't think we'd want to use it if it couldn't be for the whole thing.  Two schema languages sounds terrible.  I wouldn't be opposed to JSON Schema as an expression of the schema, we have to explicitly state that the non-normative conversion to JSON must be used to be used for validation.

> Also, stumbled into some more information on the messagepack cbor relationship.

Yeah, read through that too, definitely not warm fuzzies.

> Just some numbers of tagged messages on stackoverflow:

An interesting perspective, definitely not as popular from that POV, it's also newer.  There are a lot of implementations, so a lot of folks have spent time on it, implying that it's not just a small group pushing their idea:
http://cbor.io/impls.html

Even though it seems to have some traction, there is some risk that it will die on the vine.  Whereas Protobuf has already matured and will not be going away anytime soon.

A group that invented an alternative called ION did a comparison including MessagePack, CBOR and Protobuf:
http://tutorials.jenkov.com/ion/ion-vs-other-formats.html
http://tutorials.jenkov.com/ion/ion-performance-benchmarks.html

Those are worth reading.

**Comment by andres-h** [comment]
> Some links:
> msgpack/msgpack#129

We are referencing that thread now :D

![msgpack](https://user-images.githubusercontent.com/18348380/28284953-36a58c24-6b33-11e7-9a3e-6d031b69fe33.png)

**Comment by crotwell** [comment]
Github is being a little too clever for its own good! :(

I edited and deleted that link as I don't think we want to part of that larger conversation. 

@andres-h maybe delete the link in your quote as well?

**Comment by chad-iris** [comment]
> I edited and deleted that link as I don't think we want to part of that larger conversation.

Agreed.

**Comment by andres-h** [comment]
OK, no problem, link deleted. Let's see if the reference disappears too ;)

**Comment by andres-h** [comment]
So the guy hijacked MessagePack and named the format after himself (C. BORmann). Wow.

**Comment by chad-iris** [comment]
In the description comparing/contrasting ION with other formats at http://tutorials.jenkov.com/ion/ion-vs-other-formats.html
is this:

> ION, CBOR, MessagePack and JSON are all self describing, meaning you don't need an external schema to read them. This is essential for a network protocol where intermediate nodes may have to route messages along to other nodes. According to Protobuf's own docs you cannot see where one Protobuf message ends and the next begins, meaning Protobuf is not fully self describing. You can see where the individual Protobuf fields start and end, but not the full message.
> 
> The fact that Protobuf is not fully self describing makes it unsuitable as a network protocol message format (although you could route Protobuf messages inside other types of messages). That a data format is self describing also means that it is possible to convert a file of these formats to a textual format (JSON is alread textual) to see what is actually stored in the file.

That gives me pause about Protobuf.  I'm not sure if those are weaknesses for how we might want to use it or not, maybe @andres-h has worked around them somehow?

**Comment by andres-h** [comment]
On Tuesday 2017-07-18 23:40, Chad Trabant wrote:

>In the description comparing/contrasting ION with other formats at
>http://tutorials.jenkov.com/ion/ion-vs-other-formats.html
>is this:
>
>      ION, CBOR, MessagePack and JSON are all self describing, meaning you don't need an external
>      schema to read them. This is essential for a network protocol where intermediate nodes may
>      have to route messages along to other nodes. According to Protobuf's own docs you cannot see
>      where one Protobuf message ends and the next begins, meaning Protobuf is not fully self
>      describing. You can see where the individual Protobuf fields start and end, but not the full
>      message.
>
>      The fact that Protobuf is not fully self describing makes it unsuitable as a network protocol
>      message format (although you could route Protobuf messages inside other types of messages).
>      That a data format is self describing also means that it is possible to convert a file of
>      these formats to a textual format (JSON is alread textual) to see what is actually stored in
>      the file.
>
>That gives me pause about Protobuf. I'm not sure if those are weaknesses for how we might want to use it
>or not, maybe @andres-h has worked around them somehow?

This is exactly the strength of Protobuf!!!

This means, when you concatenate two Protobuf messages, the result is a 
Protobuf message! That's why SeedLink can transfer individual blockettes 
for low latency!

That's why we need the record header that contains record length!

This is not a theory, but works in practice as demonstrated by my Python 
and Javascript implementations!

Thanks to Protobuf, it is trivial to convert the *whole* miniSEED record 
to Javascript object (JSON), which makes handling the data in Javascript 
super easy. No more messing with DataView, CBOR, etc...

**Comment by chad-iris** [comment]
> This is exactly the strength of Protobuf!!!

OK, great, I guess; it's confusing because what you said is the opposite of what they wrote.  They implied that Protobuf was not self describing and therefore could not be converted to a textual format.  But maybe this isn't a problem when the schema is proscribed like we are doing with miniSEED.

> Thanks to Protobuf, it is trivial to convert the *whole* miniSEED record to Javascript object (JSON), which makes handling the data in Javascript super easy. No more messing with DataView, CBOR, etc...

I think that can be done for CBOR too with: https://github.com/paroga/cbor-js. What is unique to Protobuf in this regard?

**Comment by crotwell** [comment]
I'll say again, the problem with protobuf for the extra headers is that unless you have the .proto file that corresponds to those particular extra headers, all that you know is an integer field number, a byte length and an array of bytes. Protobuf does not carry with it enough information to do what I called "parsing without meaning". You can't determine even the type of an unknown field, ie float,  int, or string. Also Protobuf is simply not a key-value format because the keys are not part of the format.

Moreover, if two vendors or data centers both decide to add an extra header and both add field number 17 to their extended .proto schema file, they have created two incompatible miniseed records. The only way around this would be a global registry, and I very strongly object to creating a long term archive format that requires an dynamic external global integer registry.

Protobuf looks really nice as a over the wire format where you have significant control over both ends, which is what it was designed for, but not as a global long term archiving format.

For the stuff that must be there, protobuf is too much, we should just lay out the bytes in a fixed header. 
For the stuff that is extra, protobuf is too little, we need a self-describing key-value store.

**Comment by andres-h** [comment]
On Thursday 2017-07-20 01:29, Chad Trabant wrote:

>But maybe
>this isn't a problem when the schema is proscribed like we are doing with miniSEED.

Exactly.

>      Thanks to Protobuf, it is trivial to convert the whole miniSEED record to Javascript object
>      (JSON), which makes handling the data in Javascript super easy. No more messing with
>      DataView, CBOR, etc...
>
>I think that can be done for CBOR too with: https://github.com/paroga/cbor-js. What is unique to Protobuf
>in this regard?

1. I was talking about the whole record. You can do it with extra 
headers only, not with SimpleHeader.

Moreover:

2. With Protobuf you can transfer partial records, because each field 
(blockette) is a Protobuf message on its own.

3. With Protobuf you can append fields (blockettes) to the record 
without modifying existing data (and invalidating CRC).

4. With Protobuf you can add more fields without using extra headers as 
a hack (https://github.com/iris-edu/mseed3-evaluation/issues/21#issuecomment-313936532).

5. With Protobuf you save space, because you don't repeat the field 
names over and over again.

6. Protobuf is so simple you can implement your own parser in a day. You 
can include the encoding description in miniSEED standard like it is 
done with Steim encoding.

Would be funny to see the new miniSEED standard with 53-page CBOR RFC in 
the appendix :D

**Comment by chad-iris** [comment]
> 1. I was talking about the whole record. You can do it with extra
headers only, not with SimpleHeader.

Not unique to Protobuf.

> 3. With Protobuf you can append fields (blockettes) to the record
without modifying existing data (and invalidating CRC).

Appending data items is not unique to Protobuf.  Whether it invalidates CRC depends on where the CRC is locate and what it is calculated for.  If adding information to a record does not "invalidate" the CRC, then it was not a CRC for the whole record, which is a whole other discussion.

> 4. With Protobuf you can add more fields without using extra headers as
a hack (#21 (comment)).

Not unique to Protobuf.

> 5. With Protobuf you save space, because you don't repeat the field
names over and over again.

Not unique to Protobuf.

> 6. Protobuf is so simple you can implement your own parser in a day. You
can include the encoding description in miniSEED standard like it is
done with Steim encoding.

I agree this would be valuable.  I'm not sure it's unique to Protobuf though.

> Would be funny to see the new miniSEED standard with 53-page CBOR RFC in
the appendix :D

With an RFC it's a stable document and we can just refer to it I would think.  The CBOR RFC is quite comprehensive with lots of details beyond the encoding (mapping to/from JSON, etc, etc.).  In contrast, the Protobuf ID (not an RFC) is spartan.  Complexity is important, but page count of those two documents is a poor basis for technical evaluation.

It may not matter but the internet standards draft from Protobuf expired in 2013, there is no record of review or iteration, never made it to RFC.

@andres-h 
From what I understand from your comments above your reasoning conflates your preferred information structuring (schema) with your currently preferred encoding (Protobuf) in a way that makes it hard to separate.  Of course you can approach this however you'd like, but you should understand that this makes it more difficult for folks to understand the relative merits of the different design decisions, e.g. Protobuf.

**Comment by andres-h** [comment]
> It may not matter but the internet standards draft from Protobuf expired in 2013, there is no record of review or iteration, never made it to RFC.

Yes, I know that. Apparently Google was not interested in pushing it. Maybe the encoding is just so trivial that RFC did not make sense.

> From what I understand from your comments above your reasoning conflates your preferred information structuring (schema) with your currently preferred encoding (Protobuf) in a way that makes it hard to separate.

Really? My proposal is the so called "chunks" format aka option <span>#</span>3 in the white paper. I've suggested [two encodings](https://github.com/iris-edu/mseed3-evaluation/wiki/Chunks#encoding), one of which is Protobuf. Description of the chunks format is independent of encoding, except some remarks. I wrote most of the text before I was an advocate of Protobuf after all.

Protobuf just fits with the chunks format perfectly, but it's true that what I wrote is not unique to Protobuf. It applies to chunks format regardless of encoding. Except that you can use standard toolsets with Protobuf (record [w/o archive header] is a valid Protobuf message).

Go on with CBOR, I'm not holding you back :)

**Comment by chad-iris** [comment]
>> It may not matter but the internet standards draft from Protobuf expired in 2013, there is no record of review or iteration, never made it to RFC.

> Yes, I know that. Apparently Google was not interested in pushing it. Maybe the encoding is just so trivial that RFC did not make sense.

This is not just for you.  It is a public record of a conversation that may be visited by others wanting to understand the evaluation, merits and problems of various approaches.  Which is why...

> Protobuf just fits with the chunks format perfectly, but it's true that what I wrote is not unique to Protobuf. 

clarifying that when you write "Protobuf" you actually mean "Option 3" is important.
